kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md

```diff
@@ -104,6 +114,78 @@ Output Extraction
+
+The `callStack` cell stores a list of previous VM execution states.
+
+-   `#pushCallStack` saves a copy of VM execution state on the `callStack`.
+-   `#popCallStack` restores the top element of the `callStack`.
+-   `#dropCallStack` removes the top element of the `callStack`.
+
+```k
+    syntax InternalOp ::= "#pushCallStack"
+ // --------------------------------------
+    rule <k> #pushCallStack => .K ... </k>
+         <callStack> STACK => ListItem(<callState> CALLSTATE </callState>) STACK </callStack>
+         <callState> CALLSTATE </callState>
+
+    syntax InternalOp ::= "#popCallStack"
+ // -------------------------------------
+    rule <k> #popCallStack => .K ... </k>
+         <callStack> ListItem(<callState> CALLSTATE </callState>) REST => REST </callStack>
+         <callState> _ => CALLSTATE </callState>
+
+    syntax InternalOp ::= "#dropCallStack"
+ // --------------------------------------
+    rule <k> #dropCallStack => .K ... </k>
+         <callStack> ListItem(_) REST => REST </callStack>
+```
+
+### The StateStack
+
+- In this version that StateStack call hooks to GEth callbacks that save or restore a snapshot of the world state.
+
+-   `#pushWorldState` save the current snapshot of the world state.
+-   `#popWorldState` restores Nth snapshot of the world state.
+-   `#dropWorldState` removes the last number of snapshots of the world state. It doesn't affect the state jornal. We follow this convention as GETh never decrease the size of its journal orinally.
+
+```vlm
+    syntax InternalOp ::= "#pushWorldState"
+ // ---------------------------------------
+    rule <k> #pushWorldState => PushInsideCall() ... </k>
+         <snapshotCount> SC => SC +Int 1 </snapshotCount>
+         <snapshot>  SS => pushList(SC +Int 1, SS) </snapshot>
+
+    syntax InternalOp ::= "#popWorldState"
+ // --------------------------------------
+    rule <k> #popWorldState => RevertInsideCall(S) ... </k>
+         <snapshot> ListItem(S) SS => SS </snapshot>
+
+    syntax InternalOp ::= "#dropWorldState"
+ // ---------------------------------------
+    rule <k> #dropWorldState => .K ... </k>
+         <snapshot> ListItem(_) SS => SS </snapshot>
+```
+
+```reth
+    syntax InternalOp ::= "#pushWorldState"
+ // ---------------------------------------
+    rule <k> #pushWorldState => PushInsideCall() ... </k>
+         <snapshotCount> SC => SC +Int 1 </snapshotCount>
+         <snapshot>  SS => pushList(SC +Int 1, SS) </snapshot>
+
+    syntax InternalOp ::= "#popWorldState"
+ // --------------------------------------
+    rule <k> #popWorldState => RevertInsideCall(S) ... </k>
+         <snapshotCount> SC => SC -Int 1 </snapshotCount>
+         <snapshot> ListItem(S) SS => SS </snapshot>
+
+    syntax InternalOp ::= "#dropWorldState"
+ // ---------------------------------------
+    rule <k> #dropWorldState => Commit() ... </k>
+         <snapshotCount> SC => SC -Int 1 </snapshotCount>
+         <snapshot> ListItem(_) SS => SS </snapshot>
```

Comment:
It seems like the EVM semantics rules preserve the invariant that the snapshotCount is always the same size as the snapshot list. I don't fully understand the kllvm crate hooks, but I wonder if they preserve the same invariant. If so, it seems like we can remove the `<snapshot>` list and replace it with just the `<snapshotCount>`